src/config.zig

```diff
@@ -177,6 +171,22 @@ pub const io_depth_read = 8;
 /// The maximum number of concurrent write I/O operations to allow at once.
 pub const io_depth_write = 8;
 
+// There must be enough messages to ensure that the replica can always progress,
+// i.e. it never deadlocks because it doesn't have a free message.
+const client_table_messages_max = clients_max;
+const journal_messages_max = io_depth_read + io_depth_write;
+const loopback_queue_messages_max = 1;
+const pipelining_messages_max = pipelining_max * (1 + replicas_max);
+const quorum_messages_max = 3 * replicas_max;
+
+/// The number of full-sized messages allocated at initialization by the message bus.
+pub const message_bus_messages_max = pipelining_messages_max + quorum_messages_max +
+    loopback_queue_messages_max + client_table_messages_max + journal_messages_max;
```

Comment:
Great compendium!

I think we still need to account for outgoing messages to other replicas in the send queue?

Let's also define these constants in their own block scope to avoid introducing them in the global scope here (even as private), and then we break out the value we want to assign to `message_bus_messages_max`.

In terms of assertions, we could check that the result is greater than the number of replicas at least.

Let's also explain the magic number for `3` with a comment listing the names of all the different quorums.

To make sure we're covering all possible concurrent uses of messages, we could grep the source for all `.ref()` calls to see where we store message pointers (if you haven't already, it looks like you have!)?

Reply:
> I think we still need to account for outgoing messages to other replicas in the send queue?

Good catch, I forgot about that and the `Connection`'s `recv_message`.

> In terms of assertions, we could check that the result is greater than the number of replicas at least.

Is that misleading? Doesn't it (incorrectly) imply that values greater than `replicas_max` are safe/ok?

Maybe derived values such as `message_bus_messages_max` don't belong in `config.zig` (i.e. move to a `constants.zig` instead). The value that is currently `message_bus_messages_max` would be defined in terms of a combination other constants (e.g. `quorum_messages_max`) and configuration values (e.g. `replicas_max`).

In other words, `config.zig` should be for values that are actually configurable.

> To make sure we're covering all possible concurrent uses of messages, we could grep the source for all .ref() calls to see where we store message pointers (if you haven't already, it looks like you have!)?

I did, but evidently missed a couple so I'll double-check.

Reply:
> Is that misleading? Doesn't it (incorrectly) imply that values greater than replicas_max are safe/ok?

It's more like the assertion is saying we expect it to be at least `replicas_max`, i.e. that's as far as we can assert. In other words, we definitely wouldn't expect it to be less than the number of replicas. Alternatively, we could just assert that's greater than zero and that we didn't get a "anything times zero is zero" bug.

> In other words, config.zig should be for values that are actually configurable.

Having `constants.zig` is a good idea.

I'm thinking eventually we might have three configs: one runtime config, one cluster init time config, and one comptime config, with some runtime configs allowed to update the corresponding comptime default. We can sketch different ways of achieving this, and reorganize this as we expand our CLI options and once we have our init config being persisted in the superblock.

Reply:
> Maybe derived values such as message_bus_messages_max don't belong in config.zig (i.e. move to a constants.zig instead). The value that is currently message_bus_messages_max would be defined in terms of a combination other constants (e.g. quorum_messages_max) and configuration values (e.g. replicas_max).

Let's move the derived value into `message_bus.zig`. We've also been learning this for the LSM namespace, keeping the derived constants in there.